core/market/decentralized/src/matcher/matcher.rs

```diff
@@ -4,21 +4,60 @@ use thiserror::Error;
 use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
 
 use ya_client::model::market::{Demand, Offer, Proposal};
+use ya_client::model::ErrorMessage;
 use ya_persistence::executor::DbExecutor;
 use ya_persistence::executor::Error as DbError;
 
-use crate::protocol::{Discovery, DiscoveryBuilder, DiscoveryFactory, DiscoveryInitError};
+use crate::db::dao::*;
+use crate::db::models::Demand as ModelDemand;
+use crate::db::models::Offer as ModelOffer;
+use crate::db::*;
+use crate::migrations;
+use crate::protocol::{
+    Discovery, DiscoveryBuilder, DiscoveryError, DiscoveryFactory, DiscoveryInitError,
+};
 use crate::protocol::{OfferReceived, RetrieveOffers};
 
 #[derive(Error, Debug)]
-pub enum MatcherError {}
+pub enum DemandError {
+    #[error("Failed to insert Demand. Error: {}.", .0)]
+    InsertDemandFailure(#[from] DbError),
```

Comment:
I feel like this error (insert/remove/not-found) belongs rather to DAO than here, isn't it?  

Secondly, having `#[from]` you can use plain `?` without need of mapping error (see [line 142](https://github.com/golemfactory/yagna/pull/269/files#diff-c373b03f716ab716b09699faa1c50599R142)), but you can consider adding here `subscription_id` (see my suggestion below)

And lastly, change other error msgs below to use short syntax `{0}` also

```suggestion
    #[error("Failed to insert Demand[{1}] : {0}")]
    InsertDemandFailure(DbError, String),
```

Reply:
1. Database errors are rather ugly. I want to add more descriptive message in matcher.
2. You need error mapping when you have multiple enum variants that can be created from one error type.
3. `Subscription_id` doesn't make sense in case of insert. It was generated while ago and doesn't give any value for provider agent, that never saw this id before and will never see it again since adding Offer failed.

I changed message syntax to `{0}` as you requested in whole matcher